### ISSUE TITLE
lstm_test failure - unicharset size questions

### ISSUE THREAD

**Issue by Shreeshrii** [tesseract-ocr/tesseract#2308]
$   tesseract -v
tesseract 4.1.0-rc1-63-gb459
 leptonica-1.76.0
  libgif 5.1.4 : libjpeg 8d (libjpeg-turbo 1.4.2) : libpng 1.2.54 : libtiff 4.0.6 : zlib 1.2.8 : libwebp 0.4.4 : libopenjp2 2.3.0

```
[ RUN      ] LSTMTrainerTest.TestLayerAccess
Config file is optional, continuing...
Failed to read data from: /home/ubuntu/langdata_lstm//eng/eng.config
Warning: given outputs 1 not equal to unicharset of 113.
Num outputs,weights in Series:
  1,32,0,1:1, 0
Num outputs,weights in Series:
  C5,5:25, 0
  Ft16:16, 416
Total weights = 416
  [C5,5Ft16]:16, 416
  Mp2,2:16, 0
  Lfys32:32, 6272
  Lbx128:256, 164864
  Fc113:113, 29041
Total weights = 200593
Built network:[1,32,0,1[C5,5Ft16]Mp2,2Lfys32Lbx128Fc113] from request [1,32,0,1 Ct5,5,16 Mp2,2 Lfys32 Lbx128 O1c1]
Training parameters:
  Debug interval = 0, weights = 0.1, learning rate = 0.0005, momentum=0.9
null char=2
Loaded 520/520 pages (1-520) of document /home/ubuntu/tesseract/test/testdata/eng.Arial.exp0.lstmf
[INFO]  Setup network:SQU-lstm
lstm_test.cc:217: Failure
Expected equality of these values:
  kNumWeights[i]
    Which is: 28784
  trainer_->GetLayer(layers[i])->num_weights()
    Which is: 29041
[  FAILED  ] LSTMTrainerTest.TestLayerAccess (1009 ms)
[----------] 11 tests from LSTMTrainerTest (1172280 ms total)
```

```
  // Expected number of weights.
  const int kNumWeights[kNumLayers] = {0,
                                       0,
                                       16 * (25 + 1),
                                       0,
                                       32 * (4 * (32 + 16 + 1)),
                                       128 * (4 * (128 + 32 + 1)),
                                       128 * (4 * (128 + 32 + 1)),
                                       112 * (2 * 128 + 1)};
```

   112 * (2 * 128 + 1)}; = 28784

  113 * (2 * 128 + 1)}; = 29041

**Comment by Shreeshrii** [comment]
A minimal example shows that the size of unicharset as reported during `tesstrain.sh` and then later during `lstmtraining` are off by 1 - in case of `eng`. This could be related to the compressed unicharset being used by lstmtraining.

```
src/training/tesstrain.sh \
--fonts_dir ~/.fonts \
--lang eng \
--linedata_only   \
--noextract_font_properties \
--workspace_dir ~/tmp \
--exposures "0" \
--langdata_dir ~/langdata_lstm  \
--tessdata_dir ./tessdata \
--output_dir ~/tesstutorial/testdata \
--fontlist "Arial" "Arial Unicode MS" \
--training_text ~/langdata/eng/eng.training_text \
--maxpages 0 

rm -rf ~/tesstutorial/testdatalayer
mkdir ~/tesstutorial/testdatalayer
src/training/lstmtraining \
  --sequential_training \
  --debug_interval  0 \
  --model_output ~/tesstutorial/testdatalayer/layer \
  --continue_from ~/tessdata_best/eng.lstm \
  --append_index 5 --net_spec '[Lfx192 O1c1]' \
  --learning_rate 20e-4 \
   --traineddata ~/tesseract/test/testdata/eng/eng.traineddata \
   --traineddata ~/tesstutorial/testdata/eng/eng.traineddata \
  --train_listfile ~/tesstutorial/testdata/eng.training_files.txt \
  --max_iterations 100
 ```

**Comment by Shreeshrii** [comment]
ubuntu@tesseract-ocr:~/tesseract$ 
>  src/training/tesstrain.sh \
> --fonts_dir ~/.fonts \
> --lang eng \
> --linedata_only   \
> --noextract_font_properties \
> --workspace_dir ~/tmp \
> --exposures "0" \
> --langdata_dir ~/langdata_lstm  \
> --tessdata_dir ./tessdata \
> --output_dir ~/tesstutorial/testdata \
> --fontlist "Arial" "Arial Unicode MS" \
> --training_text ~/langdata/eng/eng.training_text \
> --maxpages 0

=== Starting training for language 'eng'
[Mon Mar 11 04:49:23 UTC 2019] /usr/local/bin/text2image --fonts_dir=/home/ubuntu/.fonts --font=Arial --outputbase=/home/ubuntu/tmp/fc-cache/sample_text.txt --text=/home/ubuntu/tmp/fc-cache/sample_text.txt --fontconfig_tmpdir=/home/ubuntu/tmp/fc-cache
Rendered page 0 to file /home/ubuntu/tmp/fc-cache/sample_text.txt.tif

=== Phase I: Generating training images ===
Rendering using Arial
[Mon Mar 11 04:49:25 UTC 2019] /usr/local/bin/text2image --fontconfig_tmpdir=/home/ubuntu/tmp/fc-cache --fonts_dir=/home/ubuntu/.fonts --strip_unrenderable_words --leading=32 --xsize=3600 --char_spacing=0.0 --exposure=0 --outputbase=/tmp/eng-2019-03-11.CJM/eng.Arial.exp0 --max_pages=0 --font=Arial --text=/home/ubuntu/langdata/eng/eng.training_text
Rendering using Arial Unicode MS
[Mon Mar 11 04:49:26 UTC 2019] /usr/local/bin/text2image --fontconfig_tmpdir=/home/ubuntu/tmp/fc-cache --fonts_dir=/home/ubuntu/.fonts --strip_unrenderable_words --leading=32 --xsize=3600 --char_spacing=0.0 --exposure=0 --outputbase=/tmp/eng-2019-03-11.CJM/eng.Arial_Unicode_MS.exp0 --max_pages=0 --font=Arial Unicode MS --text=/home/ubuntu/langdata/eng/eng.training_text
Rendered page 0 to file /tmp/eng-2019-03-11.CJM/eng.Arial.exp0.tif
Rendered page 1 to file /tmp/eng-2019-03-11.CJM/eng.Arial.exp0.tif
Rendered page 0 to file /tmp/eng-2019-03-11.CJM/eng.Arial_Unicode_MS.exp0.tif
Rendered page 1 to file /tmp/eng-2019-03-11.CJM/eng.Arial_Unicode_MS.exp0.tif

=== Phase UP: Generating unicharset and unichar properties files ===
[Mon Mar 11 04:49:28 UTC 2019] /usr/local/bin/unicharset_extractor --output_unicharset /tmp/eng-2019-03-11.CJM/eng.unicharset --norm_mode 1 /tmp/eng-2019-03-11.CJM/eng.Arial.exp0.box /tmp/eng-2019-03-11.CJM/eng.Arial_Unicode_MS.exp0.box
Extracting unicharset from box file /tmp/eng-2019-03-11.CJM/eng.Arial.exp0.box
Extracting unicharset from box file /tmp/eng-2019-03-11.CJM/eng.Arial_Unicode_MS.exp0.box
Other case É of é is not in unicharset
Wrote unicharset file /tmp/eng-2019-03-11.CJM/eng.unicharset
[Mon Mar 11 04:49:28 UTC 2019] /usr/local/bin/set_unicharset_properties -U /tmp/eng-2019-03-11.CJM/eng.unicharset -O /tmp/eng-2019-03-11.CJM/eng.unicharset -X /tmp/eng-2019-03-11.CJM/eng.xheights --script_dir=/home/ubuntu/langdata_lstm
**Loaded unicharset of size 111 from file /tmp/eng-2019-03-11.CJM/eng.unicharset**
Setting unichar properties
Other case É of é is not in unicharset
Setting script properties
Warning: properties incomplete for index 25 = ~
Writing unicharset to file /tmp/eng-2019-03-11.CJM/eng.unicharset

=== Phase E: Generating lstmf files ===
Using TESSDATA_PREFIX=./tessdata
[Mon Mar 11 04:49:28 UTC 2019] /usr/local/bin/tesseract /tmp/eng-2019-03-11.CJM/eng.Arial_Unicode_MS.exp0.tif /tmp/eng-2019-03-11.CJM/eng.Arial_Unicode_MS.exp0 --psm 6 src/training/../../tessdata/configs/lstm.train
[Mon Mar 11 04:49:28 UTC 2019] /usr/local/bin/tesseract /tmp/eng-2019-03-11.CJM/eng.Arial.exp0.tif /tmp/eng-2019-03-11.CJM/eng.Arial.exp0 --psm 6 src/training/../../tessdata/configs/lstm.train
Tesseract Open Source OCR Engine v4.1.0-rc1-63-gb459 with Leptonica
Tesseract Open Source OCR Engine v4.1.0-rc1-63-gb459 with Leptonica
Page 1
Page 1
Page 2
Page 2
Loaded 46/46 pages (1-46) of document /tmp/eng-2019-03-11.CJM/eng.Arial_Unicode_MS.exp0.lstmf
Loaded 52/52 pages (1-52) of document /tmp/eng-2019-03-11.CJM/eng.Arial.exp0.lstmf

=== Constructing LSTM training data ===
[Mon Mar 11 04:49:30 UTC 2019] /usr/local/bin/combine_lang_model --input_unicharset /tmp/eng-2019-03-11.CJM/eng.unicharset --script_dir /home/ubuntu/langdata_lstm --words /home/ubuntu/langdata_lstm/eng/eng.wordlist --numbers /home/ubuntu/langdata_lstm/eng/eng.numbers --puncs /home/ubuntu/langdata_lstm/eng/eng.punc --output_dir /home/ubuntu/tesstutorial/testdata --lang eng
**Loaded unicharset of size 111 from file /tmp/eng-2019-03-11.CJM/eng.unicharset**
Setting unichar properties
Other case É of é is not in unicharset
Setting script properties
Warning: properties incomplete for index 25 = ~
Config file is optional, continuing...
Failed to read data from: /home/ubuntu/langdata_lstm/eng/eng.config
Null char=2
Reducing Trie to SquishedDawg
Reducing Trie to SquishedDawg
Reducing Trie to SquishedDawg

=== Moving lstmf files for training data ===
Moving /tmp/eng-2019-03-11.CJM/eng.Arial.exp0.lstmf to /home/ubuntu/tesstutorial/testdata
Moving /tmp/eng-2019-03-11.CJM/eng.Arial_Unicode_MS.exp0.lstmf to /home/ubuntu/tesstutorial/testdata

Created starter traineddata for LSTM training of language 'eng'


Run 'lstmtraining' command to continue LSTM training for language 'eng'

ubuntu@tesseract-ocr:~/tesseract$
ubuntu@tesseract-ocr:~/tesseract$ rm -rf ~/tesstutorial/testdatalayer
ubuntu@tesseract-ocr:~/tesseract$ mkdir ~/tesstutorial/testdatalayer
ubuntu@tesseract-ocr:~/tesseract$ src/training/lstmtraining \
>   --sequential_training \
>   --debug_interval  0 \
>   --model_output ~/tesstutorial/testdatalayer/layer \
>   --continue_from ~/tessdata_best/eng.lstm \
>   --append_index 5 --net_spec '[Lfx192 O1c1]' \
>   --learning_rate 20e-4 \
>    --traineddata ~/tesseract/test/testdata/eng/eng.traineddata \
>    --traineddata ~/tesstutorial/testdata/eng/eng.traineddata \
>   --train_listfile ~/tesstutorial/testdata/eng.training_files.txt \
>   --max_iterations 100
Loaded file /home/ubuntu/tessdata_best/eng.lstm, unpacking...
Warning: LSTMTrainer deserialized an LSTMRecognizer!
Continuing from /home/ubuntu/tessdata_best/eng.lstm
**Appending a new network to an old one!!Warning: given outputs 1 not equal to unicharset of 110.**
Num outputs,weights in Series:
  Lfx192:192, 221952
  Fc110:110, 21230
Total weights = 243182
Built network:[1,36,0,1[C3,3Ft16]Mp3,3Lfys64Lfx96Lrx96Lfx192Fc110] from request [Lfx192 O1c1]
Training parameters:
  Debug interval = 0, weights = 0.1, learning rate = 0.002, momentum=0.5
null char=109
Loaded 72/72 pages (1-72) of document /home/ubuntu/tesstutorial/testdata/eng.Arial.exp0.lstmf
Loaded 72/72 pages (1-72) of document /home/ubuntu/tesstutorial/testdata/eng.Arial_Unicode_MS.exp0.lstmf
At iteration 100/100/100, Mean rms=5.967%, delta=50.003%, char train=102.91%, word train=99.947%, skip ratio=0%,  New worst char error = 102.91 wrote checkpoint.

Finished! Error rate = 100

**Comment by Shreeshrii** [comment]
Since the `testdata` has been generated with the current tesseract code (since original testfiles from Google are not available) a fix for lstm_test failure would be to use the currently reported weight calculation using currently reported unicharset size.

However, there is also the question of different unicharset sizes being reported by the different programs.

**Comment by Shreeshrii** [comment]
Rebuilt tesseract from master branch and recreated testdata using more pages of training_text. lstm_test passes now.

Fixed via https://github.com/tesseract-ocr/test/pull/19
